proving.go

```diff
@@ -0,0 +1,260 @@
+package merkle
+
+import (
+	"errors"
+	"math"
+)
+
+const NodeSize = 32
+
+type NodeReader interface {
+	Seek(index uint64) error
+	ReadNext() ([]byte, error)
+	Width() uint64
+}
+
+func GenerateProof(
+	provenLeafIndices []uint64,
+	readers map[uint]NodeReader,
+	hash HashFunc,
+) ([][]byte, error) {
+
+	var currentPos, subtreeStart position
+	var width uint64
+	var proof, additionalProof [][]byte
+
+	provenLeafIndexIt := &positionsIterator{s: provenLeafIndices}
+	skipPositions := &positionsStack{}
+	rootHeight := rootHeightFromWidth(readers[0].Width())
+
+	cache, err := NewTreeCache(readers, hash)
```

Comment:
Why do you still need to hold the readers map? Why not have the `TreeCache` as the input type of the cache without using the map directly (you can add a method to `TreeCache` that returns a `NodeReader` for a given height)

Reply:
I don't really "hold" it, it's just in the local scope. I did *use* it, however, which is essentially the same. I now extracted this logic into the cache, as well (`calcProofForNextLeaf`).